File: web/src/utils/file_preview.js
const MARKDOWN_EXTENSIONS = new Set(['.md', '.markdown', '.mdx'])
const IMAGE_EXTENSIONS = new Set([
  '.apng',
  '.avif',
  '.png',
  '.jpg',
  '.jpeg',
  '.gif',
  '.bmp',
  '.webp',
  '.svg'
])
const PDF_EXTENSIONS = new Set(['.pdf'])
const HTML_EXTENSIONS = new Set(['.html', '.htm'])
// 需与后端 backend/package/yuxi/services/file_preview.py 的 _OFFICE_EXTENSIONS 保持一致
const OFFICE_EXTENSIONS = new Set([
  '.docx',
  '.doc',
  '.pptx',
  '.ppt',
  '.xlsx',
  '.xls',
  '.odt',
  '.ods',
  '.odp'
])
// preset-engineering 覆盖的格式（XMind / CAD / 3D / 压缩包 / PSD / Geo / Typst / EDA / Drawing）。
// 与 office 共用"原样吐字节 + FileViewer 渲染"通道，后端 detect_preview_type 命中后会
// 返回 preview_type="office"（见 backend/_ENGINEERING_EXTENSIONS）。
const ENGINEERING_EXTENSIONS = new Set([
  // XMind
  '.xmind',
  // 压缩包
  '.zip', '.zipx', '.7z', '.rar', '.tar', '.gz', '.gzip', '.tgz',
  '.bz2', '.bzip2', '.tbz', '.tbz2', '.xz', '.txz', '.lzma', '.zst',
  '.tzst', '.cab', '.ar', '.cpio', '.iso', '.xar', '.lha', '.lzh',
  '.jar', '.war', '.ear', '.apk', '.cbz', '.cbr',
  // CAD
  '.dxf', '.dwg', '.dwf', '.dwfx', '.xps',
  // 3D 模型
  '.glb', '.gltf', '.obj', '.stl', '.ply', '.fbx', '.dae', '.3ds',
  '.3mf', '.amf', '.usd', '.usda', '.usdc', '.usdz', '.kmz',
  '.step', '.stp', '.iges', '.igs', '.ifc', '.3dm', '.brep',
  '.pcd', '.wrl', '.vrml', '.xyz', '.vtk', '.vtp',
  // Geo
  '.geojson', '.kml', '.gpx', '.shp',
  // Typst
  '.typ', '.typst',
  // EDA
  '.olb', '.dra', '.gds', '.oas', '.oasis',
  // Drawing
  '.excalidraw', '.drawio', '.dio', '.mermaid', '.mmd', '.plantuml', '.puml',
  // PSD
  '.psd', '.psb'
])
const TEXT_EXTENSIONS = new Set([
  '.txt',
  '.text',
  '.log',
  '.json',
  '.jsonl',
  '.yaml',
  '.yml',
  '.toml',
  '.ini',
  '.cfg',
  '.conf',
  '.csv',
  '.tsv',
  '.py',
  '.js',
  '.ts',
  '.jsx',
  '.tsx',
  '.vue',
  '.html',
  '.htm',
  '.css',
  '.less',
  '.scss',
  '.xml',
  '.sql',
  '.sh',
  '.bash',
  '.zsh',
  '.fish',
  '.env',
  '.dockerfile',
  '.gitignore'
])
const CODE_LANGUAGE_ALIASES = {
  js: 'javascript',
  ts: 'typescript',
  py: 'python',
  sh: 'bash',
  shell: 'bash',
  yml: 'yaml',
  docker: 'dockerfile'
}

export const normalizeCodeLanguage = (lang) => {
  const language = String(lang || '')
    .trim()
    .split(/[\s:,]/)[0]
    .toLowerCase()

  return CODE_LANGUAGE_ALIASES[language] || language
}

const CODE_LANGUAGE_MAP = {
  '.py': 'python',
  '.js': 'javascript',
  '.mjs': 'javascript',
  '.cjs': 'javascript',
  '.ts': 'typescript',
  '.tsx': 'tsx',
  '.jsx': 'jsx',
  '.vue': 'xml',
  '.html': 'xml',
  '.htm': 'xml',
  '.xml': 'xml',
  '.css': 'css',
  '.less': 'less',
  '.scss': 'scss',
  '.json': 'json',
  '.yaml': 'yaml',
  '.yml': 'yaml',
  '.toml': 'ini',
  '.ini': 'ini',
  '.cfg': 'ini',
  '.conf': 'ini',
  '.sh': 'bash',
  '.bash': 'bash',
  '.zsh': 'bash',
  '.fish': 'bash',
  '.sql': 'sql',
  '.java': 'java',
  '.kt': 'kotlin',
  '.go': 'go',
  '.rs': 'rust',
  '.php': 'php',
  '.rb': 'ruby',
  '.c': 'c',
  '.h': 'c',
  '.cpp': 'cpp',
  '.cc': 'cpp',
  '.cxx': 'cpp',
  '.hpp': 'cpp',
  '.cs': 'csharp',
  '.swift': 'swift',
  '.dockerfile': 'dockerfile'
}

export const getPreviewFileExtension = (path) => {
  const normalizedPath = String(path || '')
    .trim()
    .toLowerCase()
  if (!normalizedPath) return ''

  const fileName = normalizedPath.split('/').pop() || ''
  const dotIndex = fileName.lastIndexOf('.')
  if (dotIndex <= 0) return ''
  return fileName.slice(dotIndex)
}

export const isMarkdownPreview = (path, previewType) => {
  if (previewType === 'markdown') return true
  return MARKDOWN_EXTENSIONS.has(getPreviewFileExtension(path))
}

export const getPreviewTypeByPath = (path) => {
  const extension = getPreviewFileExtension(path)
  if (IMAGE_EXTENSIONS.has(extension)) return 'image'
  if (PDF_EXTENSIONS.has(extension)) return 'pdf'
  if (MARKDOWN_EXTENSIONS.has(extension)) return 'markdown'
  if (HTML_EXTENSIONS.has(extension)) return 'html'
  if (OFFICE_EXTENSIONS.has(extension) || ENGINEERING_EXTENSIONS.has(extension)) return 'office'
  if (TEXT_EXTENSIONS.has(extension)) return 'text'
  return 'unsupported'
}

export const getCodeLanguageByPath = (path) =>
  normalizeCodeLanguage(CODE_LANGUAGE_MAP[getPreviewFileExtension(path)] || '')

export const isHtmlPreview = (path) => HTML_EXTENSIONS.has(getPreviewFileExtension(path))

export const getPreviewTypeByContentType = (contentType) => {
  const normalized = String(contentType || '').toLowerCase()
  if (normalized.includes('application/pdf')) return 'pdf'
  if (normalized.startsWith('image/')) return 'image'
  if (normalized.includes('text/markdown')) return 'markdown'
  if (normalized.includes('text/html')) return 'html'
  if (normalized.startsWith('text/')) return 'text'
  if (normalized.includes('application/json')) return 'json'
  // 办公文档（OOXML / ODF / RTF 等）：交给前端 File Viewer 渲染
  if (
    normalized.includes('officedocument') ||
    normalized.includes('ms-word') ||
    normalized.includes('ms-powerpoint') ||
    normalized.includes('ms-excel') ||
    normalized.includes('opendocument') ||
    normalized.includes('wordprocessingml') ||
    normalized.includes('spreadsheetml') ||
    normalized.includes('presentationml') ||
    normalized === 'application/rtf' ||
    normalized.includes('application/msword') ||
    normalized.includes('application/vnd.ms-')
  ) {
    return 'office'
  }
  return 'unsupported'
}

export const isOfficePreview = (path) => {
  const extension = getPreviewFileExtension(path)
  // 名称沿用 isOfficePreview，但同时覆盖 preset-engineering 格式：
  // 后端对二者都返回 preview_type="office" 并原样透传字节，
  // 前端 FileViewer 合并了 office + engineering preset 后都能渲染。
  return OFFICE_EXTENSIONS.has(extension) || ENGINEERING_EXTENSIONS.has(extension)
}

export const normalizePreviewResponse = async (response, baseFile = {}) => {
  const contentType = response?.headers?.get?.('content-type') || ''

  if (contentType.includes('application/json')) {
    const payload = await response.json()
    const previewType = payload.preview_type || payload.previewType || payload.kind || 'text'
    return {
      ...baseFile,
      ...payload,
      content: payload.content ?? '',
      previewType,
      supported: payload.supported !== false,
      message: payload.message || '',
      previewUrl: ''
    }
  }

  const previewType =
    response?.headers?.get?.('x-yuxi-preview-type') || getPreviewTypeByContentType(contentType)
  const blob = await response.blob()

  return {
    ...baseFile,
    content: null,
    previewType,
    supported: previewType !== 'unsupported',
    message: previewType === 'unsupported' ? '当前文件暂不支持预览，请下载后查看' : '',
    // 办公文档由 File Viewer 直接渲染原始字节，需要 Blob 而非 object URL
    rawBlob: blob,
    previewUrl: window.URL.createObjectURL(blob)
  }
}

// 从 Content-Disposition 响应头解析真实文件名。
// File Viewer 纯靠 filename 扩展名识别格式，而知识库列表 entry 的 name 可能不带扩展名，
// 必须从下载接口的 Content-Disposition（后端返回的上传原始文件名）取真实名，否则误判为 unsupported。
// 兼容 RFC 5987 写法（filename*=UTF-8''xxx.docx）与退化写法（filename=xxx.docx）。
export const parseContentDispositionFilename = (header) => {
  if (!header) return ''
  const rfc5987Match = header.match(/filename\*\s*=\s*[^']*''\s*([^;]+)/i)
  if (rfc5987Match && rfc5987Match[1]) {
    try {
      return decodeURIComponent(rfc5987Match[1].trim())
    } catch {
      return rfc5987Match[1].trim()
    }
  }
  const legacyMatch = header.match(/filename\s*=\s*("?)([^";]+)\1/i)
  if (legacyMatch && legacyMatch[2]) {
    return legacyMatch[2].trim()
  }
  return ''
}

File: backend/package/yuxi/services/file_preview.py
from __future__ import annotations

import mimetypes
from pathlib import PurePosixPath

MAX_BINARY_PREVIEW_SIZE_BYTES = 30 * 1024 * 1024
MAX_TEXT_PREVIEW_CHARS = 250_000

_MARKDOWN_EXTENSIONS = frozenset({".md", ".markdown", ".mdx"})
_PDF_EXTENSIONS = frozenset({".pdf"})
_HTML_EXTENSIONS = frozenset({".html", ".htm"})
# 办公文档统一由前端 File Viewer 渲染，后端只负责原样吐字节，不再做 PDF 转换。
# 该集合需与前端 web/src/utils/file_preview.js 的 OFFICE_EXTENSIONS 保持一致。
_OFFICE_EXTENSIONS = frozenset(
    {".docx", ".doc", ".pptx", ".ppt", ".xlsx", ".xls", ".odt", ".ods", ".odp"}
)
# preset-engineering 覆盖的格式（XMind / CAD / 3D / 压缩包 / PSD / Geo / Typst / EDA / Drawing）。
# 复用 "office" 这条原始字节透传通道：前端 FileViewer 已在 preset 中合并 office + engineering，
# 后端无需为它们单开 preview_type，只扩展白名单避免被 ZIP/二进制签名误判为 unsupported。
# 需与前端 ENGINEERING_EXTENSIONS 保持一致。
_ENGINEERING_EXTENSIONS = frozenset(
    {
        # XMind
        ".xmind",
        # 压缩包（archive）
        ".zip", ".zipx", ".7z", ".rar", ".tar", ".gz", ".gzip", ".tgz",
        ".bz2", ".bzip2", ".tbz", ".tbz2", ".xz", ".txz", ".lzma", ".zst",
        ".tzst", ".cab", ".ar", ".cpio", ".iso", ".xar", ".lha", ".lzh",
        ".jar", ".war", ".ear", ".apk", ".cbz", ".cbr",
        # CAD
        ".dxf", ".dwg", ".dwf", ".dwfx", ".xps",
        # 3D 模型
        ".glb", ".gltf", ".obj", ".stl", ".ply", ".fbx", ".dae", ".3ds",
        ".3mf", ".amf", ".usd", ".usda", ".usdc", ".usdz", ".kmz",
        ".step", ".stp", ".iges", ".igs", ".ifc", ".3dm", ".brep",
        ".pcd", ".wrl", ".vrml", ".xyz", ".vtk", ".vtp",
        # Geo
        ".geojson", ".kml", ".gpx", ".shp",
        # Typst
        ".typ", ".typst",
        # EDA
        ".olb", ".dra", ".gds", ".oas", ".oasis",
        # Drawing
        ".excalidraw", ".drawio", ".dio", ".mermaid", ".mmd", ".plantuml", ".puml",
        # PSD
        ".psd", ".psb",
    }
)
_OFFICE_MEDIA_TYPES = {
    ".docx": "application/vnd.openxmlformats-officedocument.wordprocessingml.document",
    ".doc": "application/msword",
    ".pptx": "application/vnd.openxmlformats-officedocument.presentationml.presentation",
    ".ppt": "application/vnd.ms-powerpoint",
    ".xlsx": "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet",
    ".xls": "application/vnd.ms-excel",
    ".odt": "application/vnd.oasis.opendocument.text",
    ".ods": "application/vnd.oasis.opendocument.spreadsheet",
    ".odp": "application/vnd.oasis.opendocument.presentation",
}
_TEXT_EXTENSIONS = frozenset(
    {
        ".txt",
        ".text",
        ".log",
        ".json",
        ".jsonl",
        ".yaml",
        ".yml",
        ".toml",
        ".ini",
        ".cfg",
        ".conf",
        ".csv",
        ".tsv",
        ".py",
        ".js",
        ".ts",
        ".jsx",
        ".tsx",
        ".vue",
        ".css",
        ".less",
        ".scss",
        ".xml",
        ".sql",
        ".sh",
        ".bash",
        ".zsh",
        ".fish",
        ".env",
        ".dockerfile",
        ".gitignore",
        ".weather",
    }
)
_IMAGE_EXTENSIONS = frozenset({".png", ".jpg", ".jpeg", ".gif", ".bmp", ".webp", ".svg"})
_BINARY_SIGNATURES = (
    b"\x7fELF",
    b"MZ",
    b"%PDF-",
    b"PK\x03\x04",
    b"PK\x05\x06",
    b"PK\x07\x08",
    b"\x89PNG\r\n\x1a\n",
    b"\xff\xd8\xff",
    b"GIF87a",
    b"GIF89a",
    b"RIFF",
)


def is_binary_preview_type(preview_type: str) -> bool:
    """预览时需要原样返回字节流（而非文本内容）的类型。"""
    return preview_type in {"image", "pdf", "office"}


def render_preview_too_large_payload() -> dict:
    return {
        "content": None,
        "preview_type": "unsupported",
        "supported": False,
        "message": "文件过大，当前仅支持 30 MB 以内的文件预览",
        "truncated": False,
        "limit": MAX_BINARY_PREVIEW_SIZE_BYTES,
    }


def detect_preview_type(path: str, raw_content: bytes) -> tuple[str, bool, str | None]:
    suffix = PurePosixPath(path).suffix.lower()
    mime_type, _encoding = mimetypes.guess_type(path)
    head = raw_content[:1024]

    if suffix in _IMAGE_EXTENSIONS or (mime_type and mime_type.startswith("image/")):
        return "image", True, None

    if suffix in _PDF_EXTENSIONS or mime_type == "application/pdf" or head.startswith(b"%PDF-"):
        return "pdf", True, None

    # 办公文档原样交给前端 File Viewer；必须早于下方的 ZIP/二进制签名判定，
    # 否则 OOXML（PK\x03\x04 开头）会被误判为不支持预览。
    if suffix in _OFFICE_EXTENSIONS:
        return "office", True, None

    # preset-engineering 覆盖的格式（XMind / CAD / 3D / 压缩包 / PSD / Geo / Typst / EDA / Drawing）：
    # 同样原样交给前端 File Viewer 渲染。XMind/压缩包本质是 ZIP，必须在此早返回，
    # 否则会被下方的 ZIP 签名分支误判为 unsupported。
    if suffix in _ENGINEERING_EXTENSIONS:
        return "office", True, None

    if suffix in _MARKDOWN_EXTENSIONS:
        return "markdown", True, None

    if suffix in _HTML_EXTENSIONS:
        return "html", True, None

    if suffix in _TEXT_EXTENSIONS:
        return "text", True, None

    if b"\x00" in head:
        return "unsupported", False, "当前文件是二进制文件，暂不支持预览"

    if any(head.startswith(signature) for signature in _BINARY_SIGNATURES):
        if head.startswith(b"RIFF") and b"WEBP" in head[:16]:
            return "image", True, None
        return "unsupported", False, "当前文件格式暂不支持预览"

    if mime_type:
        if mime_type.startswith("text/"):
            return "text", True, None
        if mime_type in {"application/json", "application/xml", "application/javascript"}:
            return "text", True, None
        if mime_type.startswith("application/"):
            return "unsupported", False, "当前文件格式暂不支持预览"

    if not raw_content:
        return "text", True, None

    try:
        raw_content.decode("utf-8")
        return "text", True, None
    except UnicodeDecodeError:
        return "unsupported", False, "当前文件不是可读文本，暂不支持预览"


def render_preview_payload(path: str, raw_content: bytes) -> dict:
    preview_type, supported, message = detect_preview_type(path, raw_content)

    if is_binary_preview_type(preview_type) or not supported:
        return {
            "content": None,
            "preview_type": preview_type,
            "supported": supported,
            "message": message,
            "truncated": False,
            "limit": None,
        }

    try:
        content = raw_content.decode("utf-8")
    except UnicodeDecodeError:
        return {
            "content": None,
            "preview_type": "unsupported",
            "supported": False,
            "message": "当前文件不是 UTF-8 文本，暂不支持预览",
            "truncated": False,
            "limit": None,
        }

    truncated = len(content) > MAX_TEXT_PREVIEW_CHARS
    if truncated:
        content = content[:MAX_TEXT_PREVIEW_CHARS]

    return {
        "content": content,
        "preview_type": preview_type,
        "supported": True,
        "message": message,
        "truncated": truncated,
        "limit": MAX_TEXT_PREVIEW_CHARS,
    }


def detect_media_type(path: str, raw_content: bytes | None = None) -> str:
    """Detect response media type, preferring file signatures over filename suffixes."""
    head = (raw_content or b"")[:512]

    if head.startswith(b"\x89PNG\r\n\x1a\n"):
        return "image/png"
    if head.startswith(b"\xff\xd8\xff"):
        return "image/jpeg"
    if head.startswith(b"GIF87a") or head.startswith(b"GIF89a"):
        return "image/gif"
    if head.startswith(b"RIFF") and b"WEBP" in head[:16]:
        return "image/webp"
    if head.startswith(b"BM"):
        return "image/bmp"
    if head.startswith(b"%PDF-"):
        return "application/pdf"

    stripped_head = head.lstrip()
    if stripped_head.startswith(b"<svg") or stripped_head.startswith(b"<?xml"):
        suffix = PurePosixPath(path).suffix.lower()
        if suffix == ".svg" or b"<svg" in stripped_head[:256]:
            return "image/svg+xml"

    suffix = PurePosixPath(path).suffix.lower()
    if suffix in _OFFICE_MEDIA_TYPES:
        return _OFFICE_MEDIA_TYPES[suffix]

    return mimetypes.guess_type(path)[0] or "application/octet-stream"
